keps/sig-apps/3939-allow-replacement-when-fully-terminated/README.md

```diff
@@ -0,0 +1,877 @@
+status would show that the number of active pods doesn't match the desired
+parallelism. To provide better visibility, the job status can have a new field
+to track the number of Pods currently terminating.
+
+This new field can also be used by queueing controllers, such as Kueue,
+to track the number of terminating pods to calculate quotas.
+
+### Goals
+
+- Job controller should allow for flexibility in waiting for pods to be fully terminated before
+  creating replacement Pods
+- Job controller will have a new status field where we include the number of terminating pods.
+
+### Non-Goals
+
+- Other workload APIs are not included in this proposal.
+
+## Proposal
+
+The Job controller gets a list of active pods.  Active pods are pods that don't
+have a terminal phase (`Succeeded` or `Failed`) and are not terminating 
+(have a `deletionTimestamp`)
+In this KEP, we will consider terminating pods to be separate from active and failed.  
+As an opt-in behavior, the job controller can use the active and terminating
+pods to determine whether replacement Pods are needed.
+
+We propose two new API fields:
+
+1. A field in Spec that allows for opt-in behavior of whether to wait for
+   terminating pods to finish before creating replacement pods.
+2. A new field in Status for tracking the number of terminating pods.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine learning user, ML frameworks allow scheduling of multiple pods.  
+The Job controller does not typically wait for terminating pods to be marked as failed.  
+Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.
+
+This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  
+See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.
+
+#### Story 2
+
+As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  
+Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
+Replacement pods do not produce unnecessary scale ups.
+
+#### Story 3
+
+As a Job-level quota controller, I want to track the number of terminating pods,
+in addition to the active pods.
+
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### The default job controller behavior
+
+Based on the [proposed API](#job-api-definition) below, the behavior of the
+job controller prior to this KEP is equivalent to
+`podReplacementPolicy: TerminatingOrFailed`.
+
+This behavior has the following semantic problems:
+- A terminating Pod might gracefully terminate as Succeeded, but it counts
+  towards `.status.failed` as soon as it's terminating.
```

Comment:
And it won't be reclassified later (if the pods actually would terminate as Succeeded), right?

Can you please clarify?